smartsim/_core/_install/utils.py

```diff
@@ -0,0 +1,161 @@
+    """
+
+    @staticmethod
+    def _from_local_archive(
+        source: _PathLike,
+        destination: pathlib.Path,
+        **kwargs: t.Any,
+    ) -> None:
+        """Decompress a local archive
+
+        :param source: Path to the archive on a local system
+        :param destination: Where to unpack the archive
+        """
+        if tarfile.is_tarfile(source):
+            tarfile.open(source).extractall(path=destination, **kwargs)
+        if zipfile.is_zipfile(source):
+            zipfile.ZipFile(source).extractall(path=destination, **kwargs)
+
+    @staticmethod
+    def _from_local_directory(
+        source: _PathLike,
+        destination: pathlib.Path,
+        **kwargs: t.Any,
+    ) -> None:
+        """Copy the contents of a directory
+
+        :param source: source directory
+        :param destination: desitnation directory
+        """
+        shutil.copytree(source, destination, **kwargs)
+
+    @classmethod
+    def _from_http(
+        cls,
+        source: str,
+        destination: pathlib.Path,
+        **kwargs: t.Any,
+    ) -> None:
+        """Download and decompress a package
+
+        :param source: URL to a particular package
+        :param destination: Where to unpack the archive
+        """
+        local_file, _ = urlretrieve(source, **kwargs)
+        cls._from_local_archive(local_file, destination)
+        os.remove(local_file)
+
+    @staticmethod
+    def _from_git(
+        source: str, destination: pathlib.Path, **clone_kwargs: t.Any
+    ) -> None:
+        """Clone a repository
+
+        :param source: Path to the remote (URL or local) repository
+        :param destination: where to clone the repository
+        :param clone_kwargs: various options to send to the clone command
+        """
+        is_mac = OperatingSystem.autodetect() == OperatingSystem.DARWIN
+        is_arm64 = Architecture.autodetect() == Architecture.ARM64
+        if is_mac and is_arm64:
+            config_options = ["--config core.autocrlf=false", "--config core.eol=lf"]
+        else:
+            config_options = None
+        git.Repo.clone_from(
+            source, destination, multi_options=config_options, **clone_kwargs
+        )
+
+    @classmethod
+    def retrieve(
+        cls, source: _PathLike, destination: pathlib.Path, **retrieve_kwargs: t.Any
+    ) -> None:
+        """Primary method for retrieval
+
+        Automatically choose the correct method based on the extension and/or source
+        of the archive. If downloaded, this will also decompress the archive and
+        extract
+
+        :param source: URL or path to find the package
+        :param destination: where to place the package
+        :raises UnsupportedArchive: Unknown archive type
+        :raises PathNotFound: Path to archive does not exist
+        """
+        url_scheme = urlparse(str(source)).scheme
+        if str(source).endswith(".git"):
+            cls._from_git(str(source), destination, **retrieve_kwargs)
```

Comment:
Consider using the `ParseResult` path directly and stripping off any trailing `/` when making this comparison:
```suggestion
        parse_res = urlparse(str(source))
        url_scheme = parse_res.scheme
        if parse_res.path.rstrip('/').endswith(".git"):
            cls._from_git(str(source), destination, **retrieve_kwargs)
```